engine/services/engine_service.cc

```diff
@@ -468,3 +479,293 @@ std::string EngineService::GetMatchedVariant(
+  const std::filesystem::path engine_dir_path = [&] {
+    if (user_defined_engine_path != nullptr) {
+      // for backward compatible
+      return std::filesystem::path(user_defined_engine_path +
+                                   GetEnginePath(ne));
+    } else {
+      return file_manager_utils::GetEnginesContainerPath() / ne /
+             selected_engine_variant->variant /
+             selected_engine_variant->version;
+    }
+  }();
+
+  if (!std::filesystem::exists(engine_dir_path)) {
+    CTL_ERR("Directory " + engine_dir_path.string() + " is not exist!");
+    return cpp::fail("Directory " + engine_dir_path.string() +
+                     " is not exist!");
+  }
+
+  CTL_INF("Engine path: " << engine_dir_path.string());
+
+  try {
+#if defined(_WIN32)
+    // TODO(?) If we only allow to load an engine at a time, the logic is simpler.
+    // We would like to support running multiple engines at the same time. Therefore,
+    // the adding/removing dll directory logic is quite complicated:
+    // 1. If llamacpp is loaded and new requested engine is tensorrt-llm:
+    // Unload the llamacpp dll directory then load the tensorrt-llm
+    // 2. If tensorrt-llm is loaded and new requested engine is llamacpp:
+    // Do nothing, llamacpp can re-use tensorrt-llm dependencies (need to be tested careful)
+    // 3. Add dll directory if met other conditions
+
+    auto add_dll = [this](const std::string& e_type, const std::string& p) {
+      auto ws = std::wstring(p.begin(), p.end());
+      if (auto cookie = AddDllDirectory(ws.c_str()); cookie != 0) {
+        LOG_INFO << "Added dll directory: " << p;
+        engines_[e_type].cookie = cookie;
+      } else {
+        LOG_WARN << "Could not add dll directory: " << p;
+      }
+    };
+
+    if (IsEngineLoaded(kLlamaRepo) && ne == kTrtLlmRepo) {
+      // Remove llamacpp dll directory
+      if (!RemoveDllDirectory(engines_[kLlamaRepo].cookie)) {
+        LOG_WARN << "Could not remove dll directory: " << kLlamaRepo;
+      } else {
+        LOG_INFO << "Removed dll directory: " << kLlamaRepo;
+      }
+
+      add_dll(ne, engine_dir_path.string());
+    } else if (IsEngineLoaded(kTrtLlmRepo) && ne == kLlamaRepo) {
+      // Do nothing
+    } else {
+      add_dll(ne, engine_dir_path.string());
+    }
+#endif
+    engines_[ne].dl =
+        std::make_unique<cortex_cpp::dylib>(engine_dir_path.string(), "engine");
+
+  } catch (const cortex_cpp::dylib::load_error& e) {
+    LOG_ERROR << "Could not load engine: " << e.what();
+    engines_.erase(ne);
+    return cpp::fail("Could not load engine " + ne + ": " + e.what());
+  }
+
+  auto func = engines_[ne].dl->get_function<EngineI*()>("get_engine");
+  engines_[ne].engine = func();
+
+  auto& en = std::get<EngineI*>(engines_[ne].engine);
+  if (ne == kLlamaRepo) {  //fix for llamacpp engine first
+    auto config = file_manager_utils::GetCortexConfig();
+    // TODO: crash issue with trantor logging destructor.
```

Comment:
Self note: unrem this part since the crash is fixed on this PR: https://github.com/janhq/cortex.llamacpp/pull/273